include/bout/petsc_interface.hxx

```diff
@@ -0,0 +1,600 @@
+                                                     : indexConverter->getMesh()->LocalNz;
+      std::transform(
+          pw.begin(), pw.end(), std::back_inserter(positions),
+          [this, &ny, &nz](ParallelTransform::PositionsAndWeights p) -> PetscInt {
+            return this->indexConverter->getGlobal(
+                ind_type(p.i * ny * nz + p.j * nz + p.k, ny, nz));
+          });
+      std::transform(pw.begin(), pw.end(), std::back_inserter(weights),
+                     [](ParallelTransform::PositionsAndWeights p) -> PetscScalar {
+                       return p.weight;
+                     });
+    }
+    return Element(matrix.get(), global1, global2, positions, weights);
+  }
+
+  void assemble() {
+    MatAssemblyBegin(*matrix, MAT_FINAL_ASSEMBLY);
+    MatAssemblyEnd(*matrix, MAT_FINAL_ASSEMBLY);
+  }
+
+  void destroy() {
+    if (*matrix != nullptr && initialised) {
+      MatDestroy(matrix.get());
+      *matrix = nullptr;
+      initialised = false;
+    }
+  }
+
+  PetscMatrix<T> yup(int index = 0) { return ynext(index + 1); }
+  PetscMatrix<T> ydown(int index = 0) { return ynext(-index - 1); }
+  PetscMatrix<T> ynext(int dir) {
+    if (std::is_same<T, FieldPerp>::value && yoffset + dir != 0) {
+      throw BoutException("Can not get ynext for FieldPerp");
+    } else {
+      PetscMatrix<T> result; // Can't use copy constructor because don't
+                             // want to duplicate the matrix
+      result.matrix = matrix;
+      result.indexConverter = indexConverter;
+      result.pt = pt;
+      result.yoffset = std::is_same<T, Field2D>::value ? 0 : yoffset + dir;
+      result.initialised = initialised;
+      return result;
+    }
+  }
+
+  /// Provides a reference to the raw PETSc Mat object.
+  Mat* getMatrixPointer() { return matrix.get(); }
+
+private:
+  std::shared_ptr<Mat> matrix = nullptr;
+  IndexerPtr indexConverter;
+  ParallelTransform* pt;
+  int yoffset = 0;
+  bool initialised = false;
+  PetscLib lib;
+};
+
+/*!
+ * Move reference to one Vec from \p first to \p second and
+ * vice versa.
+ */
+template <class T>
+void swap(PetscVector<T>& first, PetscVector<T>& second) {
+  std::swap(first.vector, second.vector);
+  std::swap(first.indexConverter, second.indexConverter);
+  std::swap(first.location, second.location);
+  std::swap(first.initialised, second.initialised);
+}
+
+/*!
+ * Move reference to one Mat from \p first to \p second and
+ * vice versa.
+ */
+template <class T>
+void swap(PetscMatrix<T>& first, PetscMatrix<T>& second) {
+  std::swap(first.matrix, second.matrix);
+  std::swap(first.indexConverter, second.indexConverter);
+  std::swap(first.pt, second.pt);
+  std::swap(first.yoffset, second.yoffset);
+  std::swap(first.initialised, second.initialised);
+}
+
+/*!
+ *  Performs matrix-multiplication on the supplied vector
+ */
+template <class T>
+PetscVector<T> operator*(PetscMatrix<T>& mat, PetscVector<T>& vec) {
```

Comment:
I renamed the getter routines as you suggested and added `const` overloads for them so that the matrix multiplication operator can take `const` arguments.